### ISSUE TITLE
[BUG] Adding same image twice to an Album shows error when selecting to view

### ISSUE THREAD

**Issue by jonk999** [RupertAvery/DiffusionToolkit#257]
**Describe the bug**
If an image is already added to an album and it gets added again, when selecting the image to view an error appears.
Error: An item with the same key has already been added. Key: 54
The item count for the album is also incorrectly displayed. It counts the duplicates as well.

**Version:**
1.7

**Tool:**
I don't believe the tool used influences the behaviour.

**To Reproduce**
1. Add an image to an album.
2. Add the image again to the same album.
3. Click on the image to show in the preview pane.

**Expected behavior**
Image will only be added to an album once even if it is added multiple times.
Error does not appear when clicking on the image if the image is added to the same album multiple times.
Count of images in album reflects the number of images and not how many times an image has been added.

**Affected Image(s)**
Happens with any image added to the same album more than once.

**Additional context**
It happened in previous versions as well, but have only just realised what the cause appears to be.

**Comment by RupertAvery** [comment]
I am unable to replicate the issue.

I've tried combinations of the following:

- Add to album by dragging image to album name 
- Add to album by right-clicking image > Add to album 

- Image count is correct
- No errors occur when clicking image

There's something definitely strange going on, because there is a unique constraint on the table that stores album-image relationships and the insert statement checks if it's already inserted and ignores if it does.

If the constraint is missing, the application should try to recreate it on startup, and will fail and crash if there are existing duplicates.

If you open the database in DB Browser, an app for viewing SQLite databases, you should see this under "Indices"

![image](https://github.com/user-attachments/assets/8c69e6b8-3335-4910-a953-6273286677e8)

If that is there, then it should basically be impossible to create duplicates.

**Comment by jonk999** [comment]
I do have an IDX_AlbumImage index, however it looks to be being created on AlbumImageBackup rather than AlbumImage as in your screenshot. I've taken a screenshot of everything in case there are others that are potentially not correct.
I notice the indexes AlbumImage_AlbumId and AlbumImage_ImageId also look to be using AlbumImageBackup.
![image](https://github.com/user-attachments/assets/778bfa12-3d60-41af-ac04-a49d3cc24e53)

If you let me know what to run to correct, I'm happy to create a backup of the DB and then run the needed commands.

**Comment by RupertAvery** [comment]
It looks like an incomplete migration. The process was supposed to copy the AlbumImage table to a backup without the duplicates, apply the index then drop the old table and rename the backup.

First you will have to delete the duplicates in AlbumImage. Make sure you have a backup of the database, just in case.

You will have to run this a 2-3 times (it should delete the highest record in each group of records)

```
DELETE FROM AlbumImage
WHERE ROWID IN (
    SELECT MAX(RowID) 
    FROM AlbumImage
    GROUP BY AlbumId, ImageId
	Having Count(ImageID) >  1
)
```

Then recreate the index
	
```
CREATE UNIQUE INDEX "IDX_AlbumImage" ON "AlbumImage" (
	"AlbumId",
	"ImageId"
);
```

You can also drop the AlbumImageBackup table as it is no longer needed.

```
DROP TABLE AlbumImageBackup;
```

Don't forget to click "Write Changes" after executing the statements to flush the changes to the DB

**Comment by jonk999** [comment]
Thank you very much. That appears to have fixed the issue. Now if I add a picture to an album it's already a member of the message saying added to album appears, but the count doesn't increase and the error doesn't appear when selecting the file again. The first script didn't updated anywhere near as many lines as I expected, so perhaps the SQL I used to look for duplicates wasn't correct. As it still reports a lot of entries, where as when I ran your script it only affected around 8 rows and I ran it a couple of times as advised until it reported no rows affected. 
This was the SQL I found when searching:
```
SELECT ImageId, COUNT(*)
FROM AlbumImage
GROUP BY ImageId
HAVING COUNT(*) > 1
```
I modified the existing IDX_AlbumImage index as I got an error when using the SQL you provided - I'm assuming as the index still existed. I also modified the AlbumImage_AlbumId and AlbumImage_ImageId indexes to use AlbumImage rather than AlbumImageBackup. I can only assume either this is correct or they could be removed altogether seeing as you advised to drop the AlbumImageBackup table.
![image](https://github.com/user-attachments/assets/5e339d3c-b1f3-47ba-bded-011a1cb45107)
